src/Core/Services/ApiService.cs

```diff
@@ -808,6 +795,21 @@ private ErrorResponse HandleWebError(Exception e)
                     var responseJsonString = await response.Content.ReadAsStringAsync();
                     responseJObject = JObject.Parse(responseJsonString);
                 }
+
+                if (authed
+                    &&
+                    (
+                        (tokenError && response.StatusCode == HttpStatusCode.BadRequest && responseJObject?["error"]?.ToString() == "invalid_grant")
+                        ||
+                        (logoutOnUnauthorized && response.StatusCode == HttpStatusCode.Unauthorized)
+                        ||
+                        response.StatusCode == HttpStatusCode.Forbidden
+                    ))
+                {
+                    await _logoutCallbackAsync(new Tuple<string, bool, bool>(null, false, true));
+                    return null;
+                }
```

Comment:
I'd split this in two:
In the same place as it was before I'd leave:
```c#
if (authed
    &&
    (
        (logoutOnUnauthorized && response.StatusCode == HttpStatusCode.Unauthorized)
        ||
        response.StatusCode == HttpStatusCode.Forbidden
    ))
{
    await _logoutCallbackAsync(new Tuple<string, bool, bool>(null, false, true));
    return null;
}
```
and in on the new place:
```c#
if (authed && tokenError 
    &&
    response.StatusCode == HttpStatusCode.BadRequest 
    && 
    responseJObject?["error"]?.ToString() == "invalid_grant")
{
    await _logoutCallbackAsync(new Tuple<string, bool, bool>(null, false, true));
    return null;
}
```
The reason behind this is that if let's say that the status code is `HttpStatusCode.Unauthorized` then it will waste time reading the response content; so splitting it like this we gain performance on those cases.